src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -281,6 +320,10 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
         }
     }
 
+    fn const_let_allowed(&self) -> bool {
+        self.tcx.features().const_let || self.mode == Mode::ConstFn
```

Comment:
To repeat my suggestion so it doesn't get lost - this could check `self.mir.has_shortcircuit_workaround` (name up to bikeshed) - to allow variables in all constants that don't use `&&`/`||`.